pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,350 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	migrationv1alpha1 "sigs.k8s.io/vsphere-csi-driver/pkg/apis/migration/v1alpha1"
+	cnsvolume "sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/volume"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	cnsconfig "sigs.k8s.io/vsphere-csi-driver/pkg/common/config"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/common"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+// VolumeMigrationService exposes interfaces to support VCP to CSI migration.
+// It will maintain internal state to map volume path to volume ID and reverse mapping.
+type VolumeMigrationService interface {
+	// GetVolumeID returns VolumeID for given VolumePath
+	// Returns an error if not able to retrieve VolumeID.
+	GetVolumeID(ctx context.Context, volumePath string) (string, error)
+
+	// GetVolumePath returns VolumePath for given VolumeID
+	// Returns an error if not able to retrieve VolumePath.
+	GetVolumePath(ctx context.Context, volumeID string) (string, error)
+
+	// DeleteVolumeInfo helps delete mapping of volumePath to VolumeID for specified volumeID
+	DeleteVolumeInfo(ctx context.Context, volumeID string) error
+}
+
+// volumeMigration holds migrated volume information and provides functionality around it.
+type volumeMigration struct {
+	// volumePath to volumeId map
+	volumePathToVolumeID sync.Map
+	// k8sClient helps operate on CnsVSphereVolumeMigration custom resource
+	k8sClient client.Client
+	// volumeManager helps connecting to vCenter to perform Volume Operations
+	volumeManager *cnsvolume.Manager
+	cnsConfig     *cnsconfig.Config
+}
+
+const (
+	// CRDName represent the name of cnsvspherevolumemigrations CRD
+	CRDName = "cnsvspherevolumemigrations.cns.vmware.com"
+	// CRDGroupName represent the group of cnsvspherevolumemigrations CRD
+	CRDGroupName = "cns.vmware.com"
+	// CRDSingular represent the singular name of cnsvspherevolumemigrations CRD
+	CRDSingular = "cnsvspherevolumemigration"
+	// CRDPlural represent the plural name of cnsvspherevolumemigrations CRD
+	CRDPlural = "cnsvspherevolumemigrations"
+)
+
+// onceForVolumeMigrationService is used for initializing the VolumeMigrationService singleton.
+var onceForVolumeMigrationService sync.Once
+
+// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
+var volumeMigrationInstance *volumeMigration
+
+// GetVolumeMigrationService returns the singleton VolumeMigrationService
+func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Manager, cnsConfig *cnsconfig.Config) (VolumeMigrationService, error) {
+	var err error
```

Comment:
This error needs to be a global variable, like `volumeMigrationInstance`. Maybe `volumeMigrationServiceErr`?

`onceForVolumeMigrationService.Do` will only run once. If any of the conditions within this loop fail then subsequent calls to `GetVolumeMigrationService` will have `err == nil` and return a null `volumeMigrationInstance`